File: bindings/js/test-fixture/procedures.ts
/**
 * Procedure hooks for the keyvalue grammar.
 *
 * Shows ProcedureArguments in action: the current node, its text, children,
 * and source position, plus dropIfEmpty on empty tails. Author-defined
 * grammar hooks arrive as `hook_<name>` — Key is annotated `@print`.
 */

import type { Node, ProcedureArguments } from "galley-js-node";

function textOf(node: Node): string {
  const bytes = node.text();
  if (bytes === null) return "";
  return Buffer.from(bytes).toString("utf-8");
}

function posOf(node: Node): [number, number] {
  return node.lineColumn() ?? [0, 0];
}

function parseU(text: string): number {
  let value = 0;
  for (const ch of text) {
    if (ch >= "0" && ch <= "9") value = value * 10 + (ch.charCodeAt(0) - 48);
  }
  return value;
}

function countPairs(node: Node): [number, number] {
  if (node.symbolName() === "Pair") {
    const text = textOf(node);
    const colon = text.indexOf(":");
    return [1, parseU(colon >= 0 ? text.slice(colon + 1) : "")];
  }
  let count = 0;
  let total = 0;
  for (const child of node) {
    const [childCount, childSum] = countPairs(child);
    count += childCount;
    total += childSum;
  }
  return [count, total];
}

function emit(line: string): void {
  process.stderr.write(`${line}\n`);
}

export function reduction(_args: ProcedureArguments): void {}

export function reduction_Key(_args: ProcedureArguments): void {}

export function reduction_PairList(_args: ProcedureArguments): void {}

export function reduction_KeyTail(args: ProcedureArguments): void {
  args.dropIfEmpty();
}

export function reduction_NumberTail(args: ProcedureArguments): void {
  args.dropIfEmpty();
}

export function reduction_PairListTail(args: ProcedureArguments): void {
  args.dropIfEmpty();
}

export function hook_print(args: ProcedureArguments): void {
  const node = args.currentNode();
  if (node === null) return;
  const [line, column] = posOf(node);
  emit(`@print "${textOf(node)}" at ${line}:${column}`);
}

export function reduction_Number(args: ProcedureArguments): void {
  const node = args.currentNode();
  if (node === null) return;
  const [line, column] = posOf(node);
  emit(`Number ${textOf(node)} at ${line}:${column}`);
  const value = Number.parseInt(textOf(node), 10);
  if (Number.isInteger(value) && value > 999) {
    args.reportSemanticError("value out of range");
  }
}

export function reduction_Pair(args: ProcedureArguments): void {
  const node = args.currentNode();
  if (node === null) return;
  const [line, column] = posOf(node);
  const text = textOf(node);
  const colon = text.indexOf(":");
  const key = colon >= 0 ? text.slice(0, colon) : text;
  const number = colon >= 0 ? text.slice(colon + 1) : "";
  emit(`Pair ${key}=${number} (${node.length} children) at ${line}:${column}`);
}

export function reduction_Document(args: ProcedureArguments): void {
  const node = args.currentNode();
  if (node === null) return;
  const [count, total] = countPairs(node);
  emit(`Document ${count} pairs, sum=${total}`);
}

File: bindings/c/test-fixture/procedures.c
/* Procedure hooks for the keyvalue grammar.
 *
 * Shows ProcedureArguments in action: the current node, its text, children,
 * and source position, plus drop_if_empty on empty tails. Author-defined
 * grammar hooks arrive as hook_<name> — Key is annotated @print.
 *
 * Tree queries go through galley_procedure_session + galley_node_*.
 */
#include <galley.h>
#include <stdio.h>
#include <string.h>

static int symbol_is(GalleySession *session, GalleyNodeAddress node, const char *want) {
    const char *data = NULL;
    size_t len = 0;
    size_t want_len = strlen(want);
    if (galley_node_symbol_name(session, node, &data, &len) != galley_ok || data == NULL)
        return 0;
    return len == want_len && memcmp(data, want, want_len) == 0;
}

static int node_text(GalleySession *session, GalleyNodeAddress node, const char **data, size_t *len) {
    *data = NULL;
    *len = 0;
    return galley_node_text(session, node, data, len) == galley_ok && *data != NULL;
}

static void node_pos(GalleySession *session, GalleyNodeAddress node, unsigned *line, unsigned *column) {
    *line = 0;
    *column = 0;
    galley_node_line_column(session, node, line, column);
}

static unsigned parse_u(const char *data, size_t len) {
    unsigned value = 0;
    for (size_t i = 0; i < len; ++i) {
        if (data[i] >= '0' && data[i] <= '9')
            value = value * 10u + (unsigned)(data[i] - '0');
    }
    return value;
}

static void count_pairs(GalleySession *session, GalleyNodeAddress node, unsigned *count, unsigned *sum) {
    if (symbol_is(session, node, "Pair")) {
        const char *text = NULL;
        size_t len = 0;
        ++*count;
        if (node_text(session, node, &text, &len)) {
            for (size_t i = 0; i < len; ++i) {
                if (text[i] == ':') {
                    *sum += parse_u(text + i + 1, len - i - 1);
                    break;
                }
            }
        }
        return;
    }
    GalleyNodeAddress child = galley_node_first_child(session, node);
    while (child != GALLEY_INVALID_NODE) {
        count_pairs(session, child, count, sum);
        child = galley_node_next_sibling(session, child);
    }
}

void reduction(void *args) {
    (void)args;
}

void reduction_KeyTail(void *args) { galley_procedure_drop_if_empty(args); }
void reduction_NumberTail(void *args) { galley_procedure_drop_if_empty(args); }
void reduction_PairListTail(void *args) { galley_procedure_drop_if_empty(args); }
void reduction_PairList(void *args) { (void)args; }
void reduction_Key(void *args) { (void)args; }

void hook_print(void *args) {
    GalleySession *session = galley_procedure_session(args);
    GalleyNodeAddress node = galley_procedure_current_node(args);
    const char *text = NULL;
    size_t len = 0;
    unsigned line = 0, column = 0;
    if (session == NULL || node == GALLEY_INVALID_NODE)
        return;
    node_pos(session, node, &line, &column);
    fputs("@print \"", stderr);
    if (node_text(session, node, &text, &len))
        fwrite(text, 1, len, stderr);
    fprintf(stderr, "\" at %u:%u\n", line, column);
    fflush(stderr);
}

void reduction_Number(void *args) {
    GalleySession *session = galley_procedure_session(args);
    GalleyNodeAddress node = galley_procedure_current_node(args);
    const char *text = NULL;
    size_t len = 0;
    unsigned line = 0, column = 0;
    if (session == NULL || node == GALLEY_INVALID_NODE)
        return;
    node_pos(session, node, &line, &column);
    fputs("Number ", stderr);
    if (node_text(session, node, &text, &len))
        fwrite(text, 1, len, stderr);
    fprintf(stderr, " at %u:%u\n", line, column);
    fflush(stderr);
    if (text != NULL && parse_u(text, len) > 999) {
        static const char message[] = "value out of range";
        galley_procedure_report_semantic_error(args, message, sizeof(message) - 1);
    }
}

void reduction_Pair(void *args) {
    GalleySession *session = galley_procedure_session(args);
    GalleyNodeAddress node = galley_procedure_current_node(args);
    const char *text = NULL;
    size_t len = 0;
    unsigned line = 0, column = 0;
    unsigned children;
    size_t colon = 0;
    if (session == NULL || node == GALLEY_INVALID_NODE)
        return;
    node_pos(session, node, &line, &column);
    children = galley_node_child_count(session, node);
    fputs("Pair ", stderr);
    if (node_text(session, node, &text, &len)) {
        while (colon < len && text[colon] != ':')
            ++colon;
        fwrite(text, 1, colon, stderr);
        fputc('=', stderr);
        if (colon < len)
            fwrite(text + colon + 1, 1, len - colon - 1, stderr);
    }
    fprintf(stderr, " (%u children) at %u:%u\n", children, line, column);
    fflush(stderr);
}

void reduction_Document(void *args) {
    GalleySession *session = galley_procedure_session(args);
    GalleyNodeAddress node = galley_procedure_current_node(args);
    unsigned count = 0, sum = 0;
    if (session == NULL || node == GALLEY_INVALID_NODE)
        return;
    count_pairs(session, node, &count, &sum);
    fprintf(stderr, "Document %u pairs, sum=%u\n", count, sum);
    fflush(stderr);
}
